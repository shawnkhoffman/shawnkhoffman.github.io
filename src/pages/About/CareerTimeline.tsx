import React from 'react';
import CareerTimelineItem from './CareerTimelineItem';

export interface CareerTimelineData {
  date: string;
  title: string;
  description: string;
  position: 'start' | 'end';
}

interface CareerTimelineProps {
  timelineData?: CareerTimelineData[];
}

const CareerTimeline: React.FC<CareerTimelineProps> = ({ timelineData }) => {
  const defaultTimelineData: CareerTimelineData[] = [
    {
      date: '2024 - Present',
      title: 'Senior Software Engineer - Crunchyroll',
      description: 'Developing next generation video streaming experience for millions of concurrent users.',
      position: 'start',
    },
    {
      date: '2021 - 2023',
      title: 'Staff Systems Reliability Engineer - Disney Studios',
      description: 'Developed cloud services and MLOps, and led end-to-end systems reliability, for Disney film production studios, StudioLab, and Disney Research.',
      position: 'end',
    },
    {
      date: '2019 - 2021',
      title: 'Lead Consultant - HPE Services',
      description: 'Led cloud platform engineering and software development projects for Fortune 500 enterprise clients.',
      position: 'start',
    },
    {
      date: '2016 - 2019',
      title: 'Senior DevOps Engineer - Rackspace',
      description: 'Built and maintained cloud software for Fortune 500 enterprise clients.',
      position: 'end',
    },
    {
      date: '2013 - 2016',
      title: 'Senior System Engineer - U.S. Department of Veterans Affairs',
      description: 'Managed cloud infrastructure and systems for VA hospitals and clinics.',
      position: 'start',
    },
    {
      date: '2012 - 2016',
      title: 'Bachelor of Science in Computer Science - University of Texas at Austin',
      description: 'Graduated with honors and completed undergraduate research on unmanned space exploration using hydrogen fuel cell technology with NASA.',
      position: 'start',
    },
    {
      date: '2005 - 2013',
      title: 'System Engineer - U.S. Army',
      description: 'Maintained secure communications systems and network infrastructure for military operations.',
      position: 'end',
    },
    {
      date: '2004 - 2005',
      title: 'System Administrator - CompUSA',
      description: 'Managed network infrastructure and system administration for retail operations.',
      position: 'start',
    },
  ];

  const dataToRender = timelineData || defaultTimelineData;

  return (
    <div className="w-full max-w-5xl mb-10 p-6 bg-base-100 rounded-lg shadow-lg">
      <h2 className="text-3xl font-semibold text-center mb-6">Career Timeline</h2>
      <ul 
        className="timeline timeline-snap-icon max-md:timeline-compact timeline-vertical"
        aria-label="Career Timeline"
      >
        {dataToRender.map((item, index) => (
          <CareerTimelineItem
            key={index}
            date={item.date}
            title={item.title}
            description={item.description}
            position={item.position}
          />
        ))}
      </ul>
    </div>
  );
};

export default CareerTimeline;
