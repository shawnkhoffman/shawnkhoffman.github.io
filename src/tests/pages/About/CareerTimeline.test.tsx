import { render, screen, within } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import CareerTimeline from '../../../pages/About/CareerTimeline';

describe('CareerTimeline', () => {
  describe('Rendering', () => {
    it('matches snapshot', () => {
      const { container } = render(<CareerTimeline />);
      expect(container).toMatchSnapshot();
    });

    it('renders timeline container with items', () => {
      render(<CareerTimeline />);
      const timeline = screen.getAllByRole('list', { name: 'Career Timeline' })[0];
      const items = within(timeline).getAllByRole('listitem');
      expect(items.length).toBeGreaterThan(0);
    });
  });

  describe('Structure', () => {
    it('maintains proper document outline', () => {
      render(<CareerTimeline />);
      const mainHeading = screen.getAllByRole('heading', { level: 2 })[0];
      const itemHeadings = screen.getAllByRole('heading', { level: 3 });
      
      expect(mainHeading).toHaveTextContent('Career Timeline');
      expect(itemHeadings.length).toBeGreaterThan(0);
    });

    it('renders each timeline item with required elements', () => {
      render(<CareerTimeline />);
      const items = screen.getAllByRole('listitem');
      
      items.forEach(item => {
        const time = within(item).getByRole('time');
        const heading = within(item).getByRole('heading', { level: 3 });
        const description = within(item).getByRole('article');
        
        expect(time).toBeInTheDocument();
        expect(heading).toBeInTheDocument();
        expect(description).toBeInTheDocument();
      });
    });
  });

  describe('Accessibility', () => {
    it('provides semantic timeline structure', () => {
      render(<CareerTimeline />);
      const timeline = screen.getAllByRole('list', { name: 'Career Timeline' })[0];
      
      expect(timeline).toHaveAttribute('aria-label', 'Career Timeline');
      expect(timeline).toHaveClass('timeline', 'timeline-vertical');
    });

    it('maintains proper content order for screen readers', () => {
      render(<CareerTimeline />);
      const items = screen.getAllByRole('listitem');
      
      items.forEach(item => {
        const time = within(item).getByRole('time');
        const heading = within(item).getByRole('heading', { level: 3 });
        const description = within(item).getByRole('article');
        
        expect(time.compareDocumentPosition(heading))
          .toBe(Node.DOCUMENT_POSITION_FOLLOWING);
        expect(heading.compareDocumentPosition(description))
          .toBe(Node.DOCUMENT_POSITION_FOLLOWING);
      });
    });

    it('ensures all content remains visible and readable', () => {
      render(<CareerTimeline />);
      const timeline = screen.getAllByRole('list', { name: 'Career Timeline' })[0];
      const items = within(timeline).getAllByRole('listitem');
      
      items.forEach(item => {
        expect(item).toBeVisible();
        expect(item).toHaveAttribute('tabindex', '0');
      });
    });
  });
});
