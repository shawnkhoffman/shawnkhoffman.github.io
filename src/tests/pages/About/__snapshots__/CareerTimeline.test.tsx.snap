// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CareerTimeline > Accessibility > provides semantic timeline structure 1`] = `
<div>
  <div
    class="w-full max-w-5xl mb-10 p-6 bg-base-100 rounded-lg shadow-lg"
  >
    <h2
      class="text-3xl font-semibold text-center mb-6"
    >
      Career Timeline
    </h2>
    <ul
      aria-label="Career Timeline"
      class="timeline timeline-snap-icon max-md:timeline-compact timeline-vertical"
    >
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-start mb-10 md:text-end"
        >
          <time
            class="font-mono italic"
          >
            2024 - Present
          </time>
          <h3
            class="text-lg font-black"
          >
            Senior Software Engineer - Crunchyroll
          </h3>
          <p
            role="article"
          >
            Developing next generation video streaming experience for millions of concurrent users.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-end mb-10 md:text-start"
        >
          <time
            class="font-mono italic"
          >
            2021 - 2023
          </time>
          <h3
            class="text-lg font-black"
          >
            Staff Systems Reliability Engineer - Disney Studios
          </h3>
          <p
            role="article"
          >
            Developed cloud services and MLOps, and led end-to-end systems reliability, for Disney film production studios, StudioLab, and Disney Research.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-start mb-10 md:text-end"
        >
          <time
            class="font-mono italic"
          >
            2019 - 2021
          </time>
          <h3
            class="text-lg font-black"
          >
            Lead Consultant - HPE Services
          </h3>
          <p
            role="article"
          >
            Led cloud platform engineering and software development projects for Fortune 500 enterprise clients.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-end mb-10 md:text-start"
        >
          <time
            class="font-mono italic"
          >
            2016 - 2019
          </time>
          <h3
            class="text-lg font-black"
          >
            Senior DevOps Engineer - Rackspace
          </h3>
          <p
            role="article"
          >
            Built and maintained cloud software for Fortune 500 enterprise clients.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-start mb-10 md:text-end"
        >
          <time
            class="font-mono italic"
          >
            2013 - 2016
          </time>
          <h3
            class="text-lg font-black"
          >
            Senior System Engineer - U.S. Department of Veterans Affairs
          </h3>
          <p
            role="article"
          >
            Managed cloud infrastructure and systems for VA hospitals and clinics.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-start mb-10 md:text-end"
        >
          <time
            class="font-mono italic"
          >
            2012 - 2016
          </time>
          <h3
            class="text-lg font-black"
          >
            Bachelor of Science in Computer Science - University of Texas at Austin
          </h3>
          <p
            role="article"
          >
            Graduated with honors and completed undergraduate research on unmanned space exploration using hydrogen fuel cell technology with NASA.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-end mb-10 md:text-start"
        >
          <time
            class="font-mono italic"
          >
            2005 - 2013
          </time>
          <h3
            class="text-lg font-black"
          >
            System Engineer - U.S. Army
          </h3>
          <p
            role="article"
          >
            Maintained secure communications systems and network infrastructure for military operations.
          </p>
        </div>
        <hr />
      </li>
      <li
        data-testid="timeline-item"
        role="listitem"
        tabindex="0"
      >
        <div
          class="timeline-middle"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            data-testid="decorative-icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="timeline-start mb-10 md:text-end"
        >
          <time
            class="font-mono italic"
          >
            2004 - 2005
          </time>
          <h3
            class="text-lg font-black"
          >
            System Administrator - CompUSA
          </h3>
          <p
            role="article"
          >
            Managed network infrastructure and system administration for retail operations.
          </p>
        </div>
        <hr />
      </li>
    </ul>
  </div>
</div>
`;
